File: crates/treeman-db/src/postgres.rs
//! Postgres driver. Native `CREATE DATABASE … TEMPLATE` for snapshot —
//! gold path among the supported engines.

use anyhow::{Context, Result};
use async_trait::async_trait;
use sqlx::Row;
use sqlx::postgres::{PgPool, PgPoolOptions};
use tracing::debug;
use treeman_core::config::PostgresConn;

use crate::{DbDriver, Engine, Namespace};

pub struct PostgresDriver {
    pool: PgPool,
}

impl PostgresDriver {
    pub async fn connect(cfg: &PostgresConn) -> Result<Self> {
        let password = cfg.password_env.as_deref()
            .map(|k| std::env::var(k).unwrap_or_default())
            .unwrap_or_default();
        let url = format!(
            "postgres://{user}:{pass}@{host}:{port}/postgres?sslmode=disable",
            user = cfg.user,
            pass = urlencode(&password),
            host = cfg.host, port = cfg.port,
        );
        let pool = PgPoolOptions::new()
            .max_connections(cfg.pool_max)
            .connect(&url).await.context("connect postgres")?;
        Ok(Self { pool })
    }

    /// Snapshot via native CREATE DATABASE … TEMPLATE. Requires:
    ///   - no active connections to template
    ///   - same owner
    ///   - same encoding/locale
    /// Steps: ALTER DATABASE template ALLOW_CONNECTIONS=false → terminate
    /// stragglers → CREATE … TEMPLATE → restore ALLOW_CONNECTIONS=true.
    pub async fn snapshot_create(&self, source: &str, template: &str) -> Result<()> {
        validate_ident(source)?;
        validate_ident(template)?;
        // ALLOW_CONNECTIONS=false to fence the template
        self.execute_outside_tx(&format!(
            r#"ALTER DATABASE "{source}" ALLOW_CONNECTIONS false"#
        )).await?;
        let res = self.snapshot_create_inner(source, template).await;
        // Restore.
        let _ = self.execute_outside_tx(&format!(
            r#"ALTER DATABASE "{source}" ALLOW_CONNECTIONS true"#
        )).await;
        res
    }

    async fn snapshot_create_inner(&self, source: &str, template: &str) -> Result<()> {
        // Terminate stragglers
        self.execute_outside_tx(&format!(
            "SELECT pg_terminate_backend(pid) FROM pg_stat_activity \
             WHERE datname = '{source}' AND pid <> pg_backend_pid()",
        )).await?;
        // Drop existing template same-name
        self.execute_outside_tx(&format!(
            r#"DROP DATABASE IF EXISTS "{template}""#
        )).await?;
        self.execute_outside_tx(&format!(
            r#"CREATE DATABASE "{template}" WITH TEMPLATE "{source}""#
        )).await?;
        debug!(source, template, "pg snapshot create");
        Ok(())
    }

    pub async fn snapshot_restore(&self, template: &str, target: &str) -> Result<()> {
        validate_ident(template)?;
        validate_ident(target)?;
        // Terminate connections to target if any.
        self.execute_outside_tx(&format!(
            "SELECT pg_terminate_backend(pid) FROM pg_stat_activity \
             WHERE datname = '{target}' AND pid <> pg_backend_pid()",
        )).await?;
        self.execute_outside_tx(&format!(
            r#"DROP DATABASE IF EXISTS "{target}""#
        )).await?;
        // Mark template not-connectable while we clone from it.
        self.execute_outside_tx(&format!(
            r#"ALTER DATABASE "{template}" ALLOW_CONNECTIONS false"#
        )).await?;
        let res = self.execute_outside_tx(&format!(
            r#"CREATE DATABASE "{target}" WITH TEMPLATE "{template}""#
        )).await;
        let _ = self.execute_outside_tx(&format!(
            r#"ALTER DATABASE "{template}" ALLOW_CONNECTIONS true"#
        )).await;
        res
    }

    /// CREATE/DROP/ALTER DATABASE must run outside any transaction; sqlx's
    /// `execute` may implicitly wrap in one for some drivers, so we use
    /// the lower-level conn handle and explicitly disable autocommit
    /// wrapping by issuing a simple query via `Executor::execute`.
    async fn execute_outside_tx(&self, sql: &str) -> Result<()> {
        sqlx::query(sql).execute(&self.pool).await
            .with_context(|| format!("pg: {sql}"))?;
        Ok(())
    }
}

#[async_trait]
impl DbDriver for PostgresDriver {
    fn engine(&self) -> Engine { Engine::Postgres }

    async fn engine_version(&self) -> Result<String> {
        let row = sqlx::query("SHOW server_version").fetch_one(&self.pool).await?;
        Ok(row.try_get::<String, _>(0)?)
    }

    async fn ensure_db(&self, name: &str) -> Result<()> {
        validate_ident(name)?;
        let exists: Option<i32> = sqlx::query_scalar(
            "SELECT 1 FROM pg_database WHERE datname = $1"
        ).bind(name).fetch_optional(&self.pool).await?;
        if exists.is_none() {
            self.execute_outside_tx(&format!(r#"CREATE DATABASE "{name}""#)).await?;
        }
        Ok(())
    }

    async fn drop_matching(&self, prefix: &str) -> Result<Vec<String>> {
        let matched = self.list_matching(prefix).await?;
        for name in &matched {
            validate_ident(name)?;
            self.execute_outside_tx(&format!(
                "SELECT pg_terminate_backend(pid) FROM pg_stat_activity \
                 WHERE datname = '{name}' AND pid <> pg_backend_pid()",
            )).await?;
            self.execute_outside_tx(&format!(r#"DROP DATABASE IF EXISTS "{name}""#)).await?;
        }
        Ok(matched)
    }

    async fn list_matching(&self, prefix: &str) -> Result<Vec<String>> {
        let like = format!("{}%", prefix.replace('_', "\\_"));
        let rows = sqlx::query("SELECT datname FROM pg_database WHERE datname LIKE $1 ESCAPE '\\'")
            .bind(&like).fetch_all(&self.pool).await?;
        Ok(rows.into_iter().filter_map(|r| r.try_get::<String, _>(0).ok()).collect())
    }

    async fn flush_namespace(&self, ns: &Namespace) -> Result<()> {
        match ns {
            Namespace::Database(name) => {
                self.drop_matching(name).await?;
                self.ensure_db(name).await?;
                Ok(())
            }
            _ => anyhow::bail!("postgres driver only supports Namespace::Database"),
        }
    }
}

fn validate_ident(s: &str) -> Result<()> {
    if s.is_empty() {
        anyhow::bail!("empty identifier");
    }
    for c in s.chars() {
        if !(c.is_ascii_alphanumeric() || c == '_') {
            anyhow::bail!("invalid pg identifier: {s:?}");
        }
    }
    Ok(())
}

fn urlencode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        if c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.' | '~') {
            out.push(c);
        } else {
            let mut buf = [0u8; 4];
            for b in c.encode_utf8(&mut buf).bytes() {
                out.push_str(&format!("%{:02X}", b));
            }
        }
    }
    out
}

File: internal/db/postgres/postgres.go
// Package postgres is the treeman PostgreSQL driver. Uses pgx via
// database/sql.
package postgres

import (
	"context"
	"database/sql"
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"

	_ "github.com/jackc/pgx/v5/stdlib"

	"golang.org/x/sync/errgroup"

	"github.com/stubbedev/treeman/internal/config"
	"github.com/stubbedev/treeman/internal/db/containerip"
	"github.com/stubbedev/treeman/internal/db/ident"
	"github.com/stubbedev/treeman/internal/db/reachability"
)

// Driver wraps a server-level *sql.DB so DDL like CREATE DATABASE
// works.
type Driver struct {
	DB  *sql.DB
	cfg config.PostgresConn
}

// Connect opens a server-level pool against `postgres` DB so we can
// issue CREATE / DROP / TEMPLATE statements at the cluster level.
// Three connection modes are auto-detected from cfg:
//
//   - Container/ComposeService set → containerip rewrites Host:Port
//     to the published mapping (or bridge-network IP fallback), then
//     dials TCP.
//   - cfg.Host starts with "/" → unix-socket DSN
//     (`postgres:///postgres?host=/var/run/postgresql`), TCP probe
//     skipped.
//   - Otherwise → plain TCP, probe-first.
func Connect(ctx context.Context, cfg config.PostgresConn) (*Driver, error) {
	if cfg.Port == 0 {
		cfg.Port = 5432
	}
	socketMode := strings.HasPrefix(cfg.Host, "/")
	if !socketMode {
		opts := containerip.Opts{
			Container:      cfg.Container,
			ComposeService: cfg.ComposeService,
			ComposeProject: cfg.ComposeProject,
			Engine:         cfg.ContainerEngine,
			Network:        cfg.Network,
			InternalPort:   cfg.Port,
		}
		if addr, err := containerip.ResolveAddr(opts); err != nil {
			return nil, fmt.Errorf("resolve container: %w", err)
		} else if addr != nil {
			cfg.Host = addr.Host
			if addr.Port != 0 {
				cfg.Port = addr.Port
			}
		}
		if err := reachability.ProbeCtx(ctx, "postgres", cfg.Host, cfg.Port); err != nil {
			if cfg.Container != "" || cfg.ComposeService != "" {
				containerip.RefreshOpts(opts)
				if addr, e := containerip.ResolveAddr(opts); e == nil && addr != nil {
					cfg.Host = addr.Host
					if addr.Port != 0 {
						cfg.Port = addr.Port
					}
					if err2 := reachability.ProbeCtx(ctx, "postgres", cfg.Host, cfg.Port); err2 != nil {
						return nil, err2
					}
				} else {
					return nil, err
				}
			} else {
				return nil, err
			}
		}
	}
	var dsn string
	if socketMode {
		dsn = fmt.Sprintf(
			"postgres://%s:%s@/postgres?sslmode=disable&host=%s",
			url.QueryEscape(cfg.User),
			url.QueryEscape(cfg.Password),
			url.QueryEscape(cfg.Host),
		)
	} else {
		dsn = fmt.Sprintf(
			"postgres://%s:%s@%s:%d/postgres?sslmode=disable",
			url.QueryEscape(cfg.User),
			url.QueryEscape(cfg.Password),
			cfg.Host, cfg.Port,
		)
	}
	db, err := sql.Open("pgx", dsn)
	if err != nil {
		return nil, fmt.Errorf("open postgres: %w", err)
	}
	configurePool(db, int(cfg.PoolMax))
	if err := db.PingContext(ctx); err != nil {
		db.Close()
		return nil, fmt.Errorf("ping postgres: %w", err)
	}
	return &Driver{DB: db, cfg: cfg}, nil
}

// configurePool — see mysql.configurePool for rationale; same
// defaults apply (database/sql's MaxIdleConns=2 default is
// pessimistic for pooled workloads, and unbounded conn lifetime
// trips errors after server restarts).
func configurePool(db *sql.DB, poolMax int) {
	if poolMax > 0 {
		db.SetMaxOpenConns(poolMax)
		idle := poolMax / 2
		if idle < 1 {
			idle = 1
		}
		db.SetMaxIdleConns(idle)
	}
	db.SetConnMaxLifetime(30 * time.Minute)
	db.SetConnMaxIdleTime(5 * time.Minute)
}

func (d *Driver) Close() error { return d.DB.Close() }

// OpenScoped returns a fresh *sql.DB scoped to `dbName`. Use when
// you need to execute SQL that hits a specific database — Postgres
// has no `USE`, so each target DB needs its own connection. The
// returned DB owns its pool; the caller MUST Close it.
func (d *Driver) OpenScoped(ctx context.Context, dbName string) (*sql.DB, error) {
	var dsn string
	if strings.HasPrefix(d.cfg.Host, "/") {
		dsn = fmt.Sprintf(
			"postgres://%s:%s@/%s?sslmode=disable&host=%s",
			url.QueryEscape(d.cfg.User),
			url.QueryEscape(d.cfg.Password),
			url.QueryEscape(dbName),
			url.QueryEscape(d.cfg.Host),
		)
	} else {
		dsn = fmt.Sprintf(
			"postgres://%s:%s@%s:%d/%s?sslmode=disable",
			url.QueryEscape(d.cfg.User),
			url.QueryEscape(d.cfg.Password),
			d.cfg.Host, d.cfg.Port,
			url.QueryEscape(dbName),
		)
	}
	db, err := sql.Open("pgx", dsn)
	if err != nil {
		return nil, fmt.Errorf("open postgres %s: %w", dbName, err)
	}
	if err := db.PingContext(ctx); err != nil {
		db.Close()
		return nil, fmt.Errorf("ping postgres %s: %w", dbName, err)
	}
	return db, nil
}

func (d *Driver) EngineVersion(ctx context.Context) (string, error) {
	var v string
	if err := d.DB.QueryRowContext(ctx, "SELECT version()").Scan(&v); err != nil {
		return "", err
	}
	return v, nil
}

// MaxConnections returns the server's max_connections GUC. SHOW
// returns text; parse it with strconv. Same role as the mysql
// driver's MaxConnections — the fanout auto-tuner reads it to size
// the outer goroutine pool against the available connection budget.
func (d *Driver) MaxConnections(ctx context.Context) (int, error) {
	var v string
	if err := d.DB.QueryRowContext(ctx, "SHOW max_connections").Scan(&v); err != nil {
		return 0, err
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return 0, fmt.Errorf("parse max_connections %q: %w", v, err)
	}
	return n, nil
}

func (d *Driver) EnsureDB(ctx context.Context, name string) error {
	qname, err := ident.QuotePostgres(name)
	if err != nil {
		return err
	}
	exists, err := d.dbExists(ctx, name)
	if err != nil {
		return err
	}
	if exists {
		return nil
	}
	_, err = d.execOutsideTx(ctx, "CREATE DATABASE "+qname)
	return err
}

func (d *Driver) DropMatching(ctx context.Context, prefix string) ([]string, error) {
	matched, err := d.ListMatching(ctx, prefix)
	if err != nil {
		return nil, err
	}
	// Parallel DROPs: each one acquires a brief AccessExclusiveLock
	// on pg_database, so the engine serializes the lock-acquisition
	// step anyway — but the pg_terminate_backend call + WAL flush per
	// drop dominate latency on a paratest fan-out (~10 dbs). Limit
	// 4 keeps us well under the typical max_connections=100 while
	// extracting most of the wall-clock win.
	g, gctx := errgroup.WithContext(ctx)
	limit := 4
	if limit > len(matched) && len(matched) > 0 {
		limit = len(matched)
	}
	if limit < 1 {
		limit = 1
	}
	g.SetLimit(limit)
	for _, n := range matched {
		n := n
		g.Go(func() error {
			qn, err := ident.QuotePostgres(n)
			if err != nil {
				return err
			}
			// Boot any straggler connections so DROP can proceed.
			_, _ = d.DB.ExecContext(gctx,
				"SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE datname = $1", n)
			if _, err := d.execOutsideTx(gctx, "DROP DATABASE IF EXISTS "+qn); err != nil {
				return fmt.Errorf("DROP DATABASE %s: %w", qn, err)
			}
			return nil
		})
	}
	if err := g.Wait(); err != nil {
		return nil, err
	}
	return matched, nil
}

func (d *Driver) ListMatching(ctx context.Context, prefix string) ([]string, error) {
	rows, err := d.DB.QueryContext(ctx,
		"SELECT datname FROM pg_database WHERE datname LIKE $1 ORDER BY datname",
		prefix+"%")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var s string
		if err := rows.Scan(&s); err != nil {
			return nil, err
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

func (d *Driver) FlushDatabase(ctx context.Context, name string) error {
	if _, err := d.DropMatching(ctx, name); err != nil {
		return err
	}
	return d.EnsureDB(ctx, name)
}

// SnapshotCreate uses native `CREATE DATABASE … TEMPLATE` — the
// fastest path Postgres offers for cloning a database.
func (d *Driver) SnapshotCreate(ctx context.Context, source, template string) error {
	qsource, err := ident.QuotePostgres(source)
	if err != nil {
		return err
	}
	qtemplate, err := ident.QuotePostgres(template)
	if err != nil {
		return err
	}
	// Fence the source so no new connections sneak in mid-clone.
	if _, err := d.execOutsideTx(ctx,
		"ALTER DATABASE "+qsource+" ALLOW_CONNECTIONS = false"); err != nil {
		return err
	}
	defer func() {
		// Use a fresh, short-lived context so the unfence runs even
		// when ctx has been cancelled or timed out mid-clone.
		// Otherwise a cancellation can leave the source DB
		// read-only-to-new-clients forever.
		bgctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()
		_, _ = d.execOutsideTx(bgctx,
			"ALTER DATABASE "+qsource+" ALLOW_CONNECTIONS = true")
	}()
	_, _ = d.DB.ExecContext(ctx,
		"SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE datname = $1", source)
	_, _ = d.execOutsideTx(ctx, "DROP DATABASE IF EXISTS "+qtemplate)
	if _, err := d.execOutsideTx(ctx,
		"CREATE DATABASE "+qtemplate+" TEMPLATE "+qsource); err != nil {
		return err
	}
	return nil
}

func (d *Driver) SnapshotRestore(ctx context.Context, template, target string) error {
	qtemplate, err := ident.QuotePostgres(template)
	if err != nil {
		return err
	}
	qtarget, err := ident.QuotePostgres(target)
	if err != nil {
		return err
	}
	if _, err := d.execOutsideTx(ctx, "DROP DATABASE IF EXISTS "+qtarget); err != nil {
		return err
	}
	_, err = d.execOutsideTx(ctx,
		"CREATE DATABASE "+qtarget+" TEMPLATE "+qtemplate)
	return err
}

func (d *Driver) DropSnapshot(ctx context.Context, template string) error {
	qtemplate, err := ident.QuotePostgres(template)
	if err != nil {
		return err
	}
	_, err = d.execOutsideTx(ctx, "DROP DATABASE IF EXISTS "+qtemplate)
	return err
}

// DatabaseExists returns true iff the named DB is present in
// pg_database. Mirrors the MySQL driver's same-named method so the
// snapshot cache hit path is engine-agnostic at the call-site.
func (d *Driver) DatabaseExists(ctx context.Context, name string) (bool, error) {
	return d.dbExists(ctx, name)
}

func (d *Driver) dbExists(ctx context.Context, name string) (bool, error) {
	var n int
	row := d.DB.QueryRowContext(ctx, "SELECT 1 FROM pg_database WHERE datname = $1", name)
	if err := row.Scan(&n); err == sql.ErrNoRows {
		return false, nil
	} else if err != nil {
		return false, err
	}
	return true, nil
}

// execOutsideTx runs a DDL statement that cannot live inside a
// transaction (CREATE / DROP / ALTER DATABASE).
func (d *Driver) execOutsideTx(ctx context.Context, stmt string) (sql.Result, error) {
	conn, err := d.DB.Conn(ctx)
	if err != nil {
		return nil, err
	}
	defer conn.Close()
	return conn.ExecContext(ctx, stmt)
}

